### PR TITLE
POS payment method support

### PR DESCRIPTION
**Pull request by rkewlani**
**Description**

**Tested scenarios**

**Fixed issue**:

### CODE REVIEW

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 1]
```diff
@@ -1038,6 +1069,86 @@ public AdyenPaymentService getAdyenPaymentService() {
         return adyenPaymentServiceFactory.createFromBaseStore(baseStoreService.getCurrentBaseStore());
     }
 
+    /**
+     * Ininiate POS Payment using Adyen Terminal API
```
Comment (Aleffio) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]:
```suggestion
     * Initiate POS Payment using Adyen Terminal API
```

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 2]
```diff
@@ -1038,6 +1069,86 @@ public AdyenPaymentService getAdyenPaymentService() {
...
+            if (StringUtils.isNotEmpty(posReceipt)) {
+                paymentsResponse.putAdditionalDataItem("pos.receipt", posReceipt);
+            }
+            return createAuthorizedOrder(paymentsResponse);
+        }
+        throw new AdyenNonAuthorizedPaymentException(terminalApiResponse);
+    }
+
+    /**
+     * Ininiate POS Payment using Adyen Terminal API
```
Comment (Aleffio) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]:
```suggestion
     * Check POS Payment Status using Adyen Terminal API
```

File: adyenv6core/src/com/adyen/v6/service/DefaultAdyenPaymentService.java [hunk 3]
```diff
@@ -361,6 +407,39 @@ public String getDeviceFingerprintUrl() {
...
+
+        LOG.debug(TerminalAPIGsonBuilder.create().toJson(terminalApiRequest));
+        TerminalAPIResponse terminalApiResponse = terminalCloudAPI.sync(terminalApiRequest);
+
+        LOG.debug(TerminalAPIGsonBuilder.create().toJson(terminalApiResponse));
+        return terminalApiResponse;
+    }
+
+    /**
+     * Send POS Payment Request using Adyen Terminal API
```
Comment (Aleffio) [inline adyenv6core/src/com/adyen/v6/service/DefaultAdyenPaymentService.java]:
```suggestion
     * Send POS Status Request using Adyen Terminal API
```

File: adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java [hunk 4]
```diff
@@ -115,4 +119,15 @@
      * Returns the Device Fingerprint url
      */
     String getDeviceFingerprintUrl();
+
+    /**
+     * Send POS Payment Request using Adyen Terminal API
+     */
+    TerminalAPIResponse sendSyncPosPaymentRequest(CartData cartData, CustomerModel customer, String serviceId) throws Exception;
+    /**
+     * Send POS Payment Request using Adyen Terminal API
```
Comment (Aleffio) [inline adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java]:
```suggestion
     * Send POS Status Request using Adyen Terminal API
```

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java [hunk 5]
```diff
@@ -95,6 +98,9 @@
     private static final String ADYEN_PAYLOAD = "payload";
     private static final String REDIRECT_RESULT = "redirectResult";
 
+    private static final int POS_TOTALTIMEOUT_DEFAULT = 130;
+    private static final String POS_TOTALTIMEOUT_KEY = "pos.totaltimeout";
```
Comment (rikterbeek) [inline adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java]: let's add it into documentation how to configure this.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java [hunk 6]
```diff
@@ -178,6 +186,42 @@ public String placeOrder(@ModelAttribute("placeOrderForm") final PlaceOrderForm
             } catch (Exception e) {
                 LOGGER.error(ExceptionUtils.getStackTrace(e));
             }
+        } else if (PAYMENT_METHOD_POS.equals(adyenPaymentMethod)) {
+            try {
+                String originalServiceId = Long.toString(System.currentTimeMillis() % 10000000000L);
+                request.setAttribute("originalServiceId", originalServiceId);
+                Long paymentStartTime = System.currentTimeMillis();
+                request.setAttribute("paymentStartTime", paymentStartTime);
+                OrderData orderData = adyenCheckoutFacade.initiatePosPayment(request, cartData);
```
Comment (rikterbeek) [inline adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java]: no timeout on the initiatePosPayment ?

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js [hunk 7]
```diff
@@ -48,6 +48,14 @@ var AdyenCheckoutHybris = (function () {
                 }
             }
 
+            if ( paymentMethod === "pos" ) {
+                var terminalId = document.getElementById('p_method_adyen_hpp_pos_terminal').value;
+                if( terminalId === "" ) {
+                    window.alert("Please select a terminal");
```
Comment (rikterbeek) [inline adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js]: can we make this text translatable ?

File: adyenv6backoffice/resources/adyenv6backoffice-backoffice-config.xml [hunk 8]
```diff
@@ -25,21 +25,28 @@
...
+					<editorArea:attribute qualifier="adyenPosEnabled"/>
+					<editorArea:attribute qualifier="adyenPosApiKey" editor="com.hybris.cockpitng.editor.defaultpassword"/>
+					<editorArea:attribute qualifier="adyenPosMerchantAccount"/>
+					<editorArea:attribute qualifier="adyenPosStoreId"/>
+					<editorArea:attribute qualifier="adyenPosRecurringContractMode"/>
+				</editorArea:section>
+				<editorArea:section name="hmc.adyen.hpp">
+					<editorArea:attribute qualifier="adyenSkinCode"/>
+					<editorArea:attribute qualifier="adyenSkinHMAC" editor="com.hybris.cockpitng.editor.defaultpassword"/>
+				</editorArea:section>
```
Comment (rikterbeek) [inline adyenv6backoffice/resources/adyenv6backoffice-backoffice-config.xml]: where do we need adyenSkinCode and adyenSkinHMAC for ?

File: adyenv6core/resources/adyenv6core-items.xml [hunk 9]
```diff
@@ -285,6 +313,10 @@
                     <description>Number of Installments</description>
                     <persistence type="property"/>
                 </attribute>
+                <attribute qualifier="adyenTerminalId" type="java.lang.String">
+                    <description>Terminal ID</description>
+                    <persistence type="property"/>
+                </attribute>
```
Comment (rikterbeek) [inline adyenv6core/resources/adyenv6core-items.xml]: we do not need this right as we are using the API to retrieve the terminals ?

File: adyenv6core/resources/localization/adyenv6core-locales_en.properties [hunk 10]
```diff
@@ -71,7 +78,13 @@ type.paymentinfo.adyenMultibancoAmount.name= Multibanco Amount
 type.paymentinfo.adyenMultibancoDeadline.name= Multibanco Deadline
 type.paymentinfo.adyenMultibancoEntity.name=Multibanco Entity
 type.paymentinfo.adyenMultibancoReference.name=Multibanco Reference
+type.PaymentInfo.adyenPosReceipt.name=POS Receipt
+type.PaymentInfo.adyenTerminalId.name=POS Terminal ID
 
+type.PaymentInfo.encryptedExpiryMonth.name = Encrypted Credit Card Expiry Month
+type.PaymentInfo.encryptedExpiryYear.name = Encrypted Credit Card Expiry Year
+type.PaymentInfo.encryptedSecurityCode.name = Encrypted Credit Card CVC
+type.PaymentInfo.encryptedCardNumber.name = Encrypted Credit Card Number
```
Comment (rikterbeek) [inline adyenv6core/resources/localization/adyenv6core-locales_en.properties]: this is unrelated to the POS payment method right did we miss this for cards ?

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 11]
```diff
@@ -1038,6 +1069,86 @@ public AdyenPaymentService getAdyenPaymentService() {
...
+                    }
+                    return createAuthorizedOrder(paymentsResponse);
+                } else {
+                    throw new AdyenNonAuthorizedPaymentException(terminalApiResponse);
+                }
+            } else {
+                ErrorConditionType errorCondition = TerminalAPIUtil.getErrorConditionForStatusFromStatusResponse(terminalApiResponse);
+                //If transaction is still in progress, keep retrying in 5 seconds.
+                if (errorCondition == ErrorConditionType.IN_PROGRESS) {
+                    TimeUnit.SECONDS.sleep(5);
```
Comment (rikterbeek) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]: why are we doing a thread sleep here ? Would avoid doing thread sleep at all times and better to do an new request from the front-end.

File: adyenv6core/src/com/adyen/v6/service/DefaultAdyenOrderService.java [hunk 12]
```diff
@@ -148,6 +148,9 @@ public void updateOrderFromPaymentsResponse(OrderModel order, PaymentsResponse p
         paymentInfo.setAdyenMultibancoDeadline(paymentsResponse.getMultibancoDeadline());
         paymentInfo.setAdyenMultibancoReference(paymentsResponse.getMultibancoReference());
 
+        //pos receipt
+        paymentInfo.setAdyenPosReceipt(paymentsResponse.getAdditionalDataByKey("pos.receipt"));
```
Comment (rikterbeek) [inline adyenv6core/src/com/adyen/v6/service/DefaultAdyenOrderService.java]: remove empty line

File: adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java [hunk 13]
```diff
@@ -115,4 +119,15 @@
...
+     */
+    TerminalAPIResponse sendSyncPosStatusRequest(CartData cartData, String serviceId) throws Exception;
+
+
```
Comment (rikterbeek) [inline adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java]: remove empty lines

File: adyenv6core/src/com/adyen/v6/util/TerminalAPIUtil.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+import org.apache.log4j.Logger;
+import com.adyen.model.nexo.DocumentQualifierType;
+import com.adyen.model.nexo.ErrorConditionType;
+import com.adyen.model.nexo.OutputText;
+import com.adyen.model.nexo.PaymentReceipt;
+import com.adyen.model.nexo.ResultType;
+import com.adyen.model.terminal.TerminalAPIResponse;
+import com.google.common.base.Splitter;
+
+public class TerminalAPIUtil {
```
Comment (rikterbeek) [inline adyenv6core/src/com/adyen/v6/util/TerminalAPIUtil.java]: any of these function important for other merchants? if so we can think of moving this into the java library.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java (hunk 6 above)
Comment (rkewlani) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java]: No, it is default connection timeout.

File: adyenv6core/src/com/adyen/v6/service/DefaultAdyenOrderService.java (hunk 12 above)
Comment (rkewlani) [reply adyenv6core/src/com/adyen/v6/service/DefaultAdyenOrderService.java]: It is intentional for formatting.

File: adyenv6core/src/com/adyen/v6/util/TerminalAPIUtil.java (hunk 14 above)
Comment (rkewlani) [reply adyenv6core/src/com/adyen/v6/util/TerminalAPIUtil.java]: As we discussed before, if we keep it in java library then we may need to provide more functions to avoid "missing utility" tickets. So we kept it in hybris.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js [hunk 15]
```diff
@@ -48,6 +48,14 @@ var AdyenCheckoutHybris = (function () {
                 }
             }
 
+            if ( paymentMethod === "pos" ) {
+                var terminalId = document.getElementById('p_method_adyen_hpp_pos_terminal').value;
```
[the remark this answers is not in the document]
Comment (rkewlani) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js]: It is consistent to keep it pos only. adyen is appended as prefix only for cards because it was present in default hybris system.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js (hunk 7 above)
Comment (rkewlani) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js]: We don't have internationalisation in java script alert messages. If we do it for this text, rest all text will also need to be changed to make all in one language.

File: adyenv6backoffice/resources/adyenv6backoffice-backoffice-config.xml (hunk 8 above)
Comment (rkewlani) [reply adyenv6backoffice/resources/adyenv6backoffice-backoffice-config.xml]: It is old code and clean up of this is not part of this story.

File: adyenv6core/resources/adyenv6core-items.xml (hunk 9 above)
Comment (rkewlani) [reply adyenv6core/resources/adyenv6core-items.xml]: This is for storing terminal information in database.

File: adyenv6core/resources/localization/adyenv6core-locales_en.properties (hunk 10 above)
Comment (rkewlani) [reply adyenv6core/resources/localization/adyenv6core-locales_en.properties]: Yes this is unrelated. It was a small bug and I fixed it because it was small change.

File: adyenv6core/src/com/adyen/v6/facades/AdyenCheckoutFacade.java [hunk 16]
```diff
@@ -203,4 +208,14 @@
     void handlePaymentForm(AdyenPaymentForm adyenPaymentForm, BindingResult bindingResult);
 
     PaymentDetailsListWsDTO getPaymentDetails(String userId) throws IOException, ApiException;
+
+    /**
+     * Ininiate POS Payment using Adyen Terminal API
```
[the remark this answers is not in the document]
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/AdyenCheckoutFacade.java]: Fixed

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 2 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]: Fixed

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (martinsrenato) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js]: Changed to 'adyen_pos_terminal'

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js [hunk 17]
```diff
@@ -90,6 +97,11 @@ var AdyenCheckoutHybris = (function () {
             if ( ssn ) {
                 $( "#socialSecurityNumber" ).val( ssn.val() );
             }
+
+            var terminalId = $( '#p_method_adyen_hpp_' + paymentMethod + '_terminal' );
```
[the remark this answers is not in the document]
Comment (martinsrenato) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js]: Changed to 'adyen_pos_terminal'

File: adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java (hunk 4 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java]: Fixed

File: adyenv6core/src/com/adyen/v6/service/DefaultAdyenPaymentService.java (hunk 3 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/service/DefaultAdyenPaymentService.java]: Fixed

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 1 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]: Fixed

File: adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java (hunk 13 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/service/AdyenPaymentService.java]: Done

File: README.md [hunk 18]
```diff
@@ -135,6 +135,12 @@ https://docs.adyen.com/developers/payment-methods/boleto-bancario/boleto-payment
 ```
 is3DS2allowed = false
 ```
+## POS Timeout configuration
+POS timeout (time calcualted since initiating a payment) is max time to keep terminal connection open. It is set to 130 seconds by default already. If you want to change it, please add following property in local.properties file, build your environment and restart the server. (Change 130 to your desired time, in seconds).
```
Comment (Aleffio) [inline README.md]:
```suggestion
POS timeout (time calculated since initiating a payment) is max time to keep terminal connection open. It is set to 130 seconds by default already. If you want to change it, please add following property in local.properties file, build your environment and restart the server. (Change 130 to your desired time, in seconds).
```

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 11 above)
Comment (rkewlani) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]: I don't see any problem in using Thread.sleep because we are not explicitly managing threads  . Thread management is being done by framework itself and we are just holding current thread for 5 seconds. Sending the call back to Front end means someone has to trigger the request manually again to check the status. We can discuss more if you want.

File: adyenv6core/src/com/adyen/v6/service/DefaultAdyenOrderService.java (hunk 12 above)
Comment (rkewlani) [reply adyenv6core/src/com/adyen/v6/service/DefaultAdyenOrderService.java]: this is kept intentional for formatting.

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 11 above)
Comment (rikterbeek) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]: Using thread sleep is in general a bad design and as well a heavy load to the server so think polling is the better solution here. You can auto poll it from the front-end so no customer interaction is needed. Hava a chat with alessio he did it for magento i believe.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js (hunk 7 above)
Comment (rikterbeek) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/webroot/_ui/responsive/common/js/adyen.checkout.js]: ok let's keep it as it is for now.

File: adyenv6core/src/com/adyen/v6/util/TerminalAPIUtil.java (hunk 14 above)
Comment (rikterbeek) [reply adyenv6core/src/com/adyen/v6/util/TerminalAPIUtil.java]: ok

File: adyenv6backoffice/resources/adyenv6backoffice-backoffice-config.xml (hunk 8 above)
Comment (rkewlani) [reply adyenv6backoffice/resources/adyenv6backoffice-backoffice-config.xml]: no it is just formatting change.

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 11 above)
Comment (rikterbeek) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]: Let's merge it for now and discuss this separately with pro/cons of this.
